### DISCUSSION TITLE
micropower/upower.py support for STM32F411

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#15828] [Libraries & Drivers] [answered]
I would like `pyb.stop()` to woken-up by a trigger signal.  C13 has some labels on it `RTC_AMP1, RTC_OUT and RTC_TS`, however C13 drives a blue LED on the Blackpill.  So, maybe I can't use that pin.

The datasheet does not show a `WUPP` bit so I guess `class WakeupPin:` is also a non-starter.

Tried changing in `class wakeup_X1:`
```
self.pin = pyb.Pin.board.X1
to
self.pin = pyb.Pin.cpu.A0
```
runs without error, but no wake-up.

**Comment by davefes** [comment]
I have verified that `PWR_CR and the CWUF bit` and `PWR_CSR and the EWUP bit` are the same as for the PYBoard v1.1.

There must be something else in `class wakeup_X1` that I have missed or there is something on the hardware platform that is different.

Any suggestions why it might not wake-up?

**Reply by davefes** [reply]
I have noticed that a 10K pull-up does not lift the low on PA0 at all.  Connecting PA0 to 3V3 does bring it up to 3V3 but still no trigger.  Also tried C13. 

Thought I'd try `machine.lightsleep()` instead of `pyb.stop()` and a 10K will pull A0 up ... but still no trigger.

I thought `machine.lightsleep` and `pyb.stop` were the same.

**Comment by peterhinch** [comment, marked as answer]
Have you seen [these notes](https://github.com/peterhinch/micropython-micropower/blob/master/UPOWER.md) on Pyboard micropower usage? It is definitely possible to wake a Pyboard from `pyb.standby` deepsleep with the alarm pins. In particular, see para 2.14 which discusses the two tamper pins.

**Reply by davefes** [reply]
Yes, I have read and am trying to just wakeup the device using pinA0.  I had a look at the tamper pin suggestion, but was not certain which were the tamper pin or pins on the STM32F411.  I focus more on that method as it seems many STM32 parts have it.

I have tried using C13 so will give `class Tamper` a try.
```py
self.pin = pyb.Pin.cpu.C13  # X18 doesn't exist on Pyboard D
```

**Reply by davefes** [reply]
Ah, C13 sits at about 1.2V and if I pull it down it wakes-up.  This looks hopeful.

Thank you.

**Reply by davefes** [reply]
1.2V is in the "Area not determined", but was HI in my case.  Put an external 10K pull-up on C13.  

How is `def _pinconfig(self):` used or when is it run?

**Reply by davefes** [reply]
Well, thanks Damien for this code snippet `machine.lightsleep()` or `pyb.stop()` now can be woken-up.  The docs do say you can use a external interrupt!

```py
p = machine.Pin('A0', machine.Pin.IN, machine.Pin.PULL_UP)
p.irq(trigger=machine.Pin.IRQ_RISING, handler=lambda p:None)
```
